code/__HELPERS/path.dm

```diff
@@ -0,0 +1,436 @@
+/**
+ * This file contains the stuff you need for using JPS (Jump Point Search) pathing, an alternative to A* that skips
+ * over large numbers of uninteresting tiles resulting in much quicker pathfinding solutions.
+ */
+
+/**
+ * This is the proc you use whenever you want to have pathfinding more complex than "try stepping towards the thing".
+ * If no path was found, returns an empty list, which is important for bots like medibots who expect an empty list rather than nothing.
+ *
+ * Arguments:
+ * * caller: The movable atom that's trying to find the path
+ * * end: What we're trying to path to. It doesn't matter if this is a turf or some other atom, we're gonna just path to the turf it's on anyway
+ * * max_distance: The maximum number of steps we can take in a given path to search (default: 30, 0 = infinite)
+ * * mintargetdistance: Minimum distance to the target before path returns, could be used to get near a target, but not right to it - for an AI mob with a gun, for example.
+ * * id: An ID card representing what access we have and what doors we can open. Its location relative to the pathing atom is irrelevant
+ * * simulated_only: Whether we consider turfs without atmos simulation (AKA do we want to ignore space)
+ * * exclude: If we want to avoid a specific turf, like if we're a mulebot who already got blocked by some turf
+ * * skip_first: Whether or not to delete the first item in the path. This would be done because the first item is the starting tile, which can break movement for some creatures.
+ * * diagonal_safety: ensures diagonal moves won't use invalid midstep turfs by splitting them into two orthogonal moves if necessary
+ */
+/proc/get_path_to(caller, end, max_distance = 30, mintargetdist, id = null, simulated_only = TRUE, turf/exclude, skip_first = TRUE, diagonal_safety = TRUE)
+	if(!caller || !get_turf(end))
+		return
+
+	var/l = SSpathfinder.mobs.getfree(caller)
+	while(!l)
+		stoplag(3)
+		l = SSpathfinder.mobs.getfree(caller)
```

Comment:
A lot of `get_path_to()` calls are syncronous, and expect this to not sleep. Are you 100% sure that nothing will attempt to call this multiple times? Say a medbot wants to path somewhere, and the server is lagging like hell, what happens if SSnpc fires twice on the same bot?

Reply:
It's hard to say honestly.

`get_path_to()` only hands out a path as a list to whatever calls it, and in the case of bots sets it as `path` and from there processes it. So if it suddently gets empty, then it'll just start over. If it's the same path but slightly behind, then either it'll correct itself (e.g. `loc == step` Alright I'm already there, next step), or if the disparity is too large, then it'll buildup `frustration` from not stepping properly and recalculate.